Improve CoqIDE documentation (#16070)
coq/coq: doc/sphinx/practical-tools/coqide.rst

If you include script file names (which end with `.v`) as arguments, each is opened
in a separate tab.  If you don't, CoqIDE opens a single unnamed buffer
(titled `*scratch*`).  `coqide` also accepts many of the options of `coqtop`
(see :ref:`thecoqcommands`), while ignoring the ones that aren't meaningful
for CoqIDE.  Use `coqide --help` to see the list of command line options.

.. _coqide_mainscreen:

  .. image:: ../_static/coqide.png
     :alt: CoqIDE main screen

The screen shot shows CoqIDE as the user is stepping through the file `Fermat.v`.
A menu bar and a tool bar appear at the top of the window. The left-hand panel shows
the current *script buffer*.  Each script buffer corresponds to a separate Coq process.
The upper right panel is the *proof panel*, which shows
the goals to be proven.

The lower right panel has three tabs: the *Messages
panel*, which shows messages produced by commands and tactics; the *Errors panel*,
which shows errors detected when running in :ref:`async mode <asyncmode>` and the
*Jobs panel,* which shows information on the worker processes used by async mode.
The contents of the right-hand panels are specific to the currently-displayed script.

> Click the arrow icons to detach these panel into separate windows.  The proof panel

Author: I think the command form reads a lot better and more easily understood than the passive voice.  And more succinct.

can be detached from the `Windows/Detach Proof` menu item.

The *status bar* is a line of text that appears at the bottom of the window.

Managing files and buffers, basic editing
-----------------------------------------

The *File* menu lets you open files, create new buffers, save buffers to a file,
and print or export them in various formats.

Text editing provides the basic operations such as copy, cut, paste, find and replace.

> Most editing operations are shown in the *Edit* menu.  Most menu entries show the equivalent

Author: "Keystroke equivalents (if defined) for menu items are shown on the right of each item."

keystrokes to perform the action.  If you need more complex editor commands, you can

> launch your favorite text editor on the current buffer, using the

Author: It was in the original text.  It was cute.  It is gone.

> launch your favorite text editor on the current buffer, using the

Reviewer: I thought it was cute too, though we cannot assume our readers will enjoy comparing text editors. 0:-)

*Edit/External Editor* menu. (Use `Edit/Preferences/Externals/External Editor` to
identify your favorite editor.)  When you're done editing, you currently must
reopen the file to see your changes.  Also note:

- Undo is also available as ``Ctrl-Z`` (`Cmd-Z` on macOS).  Redo is `Ctrl-Shift-Z`
  (`Shift-Cmd-Z` on macOS).
- Select all is `Ctrl-A` (`Cmd-A` on macOS).
- `Ctrl-Delete` deletes a word of text after the cursor
- `Ctrl-Backspace` deletes a word of text before the cursor
- Commenting and uncommenting the current line or selected text is available in
  the *Tools* menu.  If some text is selected, exactly that text is commented out;
  otherwise the line containing the cursor is commented out.  To uncomment, position

>   the cursor between `(*` and `*)` or select any text between them.

Author: I'm reluctant: a) it is on the menu and b) the user can reconfigure it.  If I just say "Ctrl-D" it could lead to puzzlement.

Files are automatically saved periodically to a recovery file.  For example,
`foo.v` is saved to `#foo.v#` every 10 seconds by default.  You can change the

> interval in the *Edit / Preferences / Files* dialog.

Author: I'd like to get rid of the recovery files to enable a smoother workflow with compile and external make.  Instead, if possible, save the files automatically when CoqIDE loses focus--though I'm not sure I can get that to work with XWindows, at least on my system.  And when CoqIDE regains focus, it should reload any files that have changed.  Users should be using a source code control system, which I recommend in #15888.

FWIW my IDE, IntelliJ IDEA, has some fabulous undo and diffing capabilities.  For example, it takes automatic snapshots of every file as you edit it.  This info is then available as "Local History".  I can edit in the right hand panel.  Clicking on ">>" undoes a change.  (I have no intention of adding anything like this to CoqIDE.)

![image](https://user-images.githubusercontent.com/1253341/172024725-6dc2d4e4-4e48-40b2-aa16-09659c5b637c.png)

Coq Integrated Development Environment
========================================

.. todo PR: how to say that a number of things are broken?  Maybe list them
   somewhere--doesn't have to be super detailed

The Coq Integrated Development Environment is a user-friendly GUI
replacement for `coqtop`. Its main purpose is to

> allow users to edit scripts and step forward and backward through Coq scripts,

Author: British English often uses "forwards", American English not so much.  For better or worse, we've chosen to use American English for the documentation (fine with me, of course, but I was not involved in the choice).

See https://www.collinsdictionary.com/dictionary/english/forward

respectively executing and undoing commands and tactics.

> To run CoqIDE, enter `coqide` in operating systems's command line.

Author: Best not to get into environment-specific details like starting coqide through shortcuts.  I expect the users who care know how to create them.

If you include the script file names as arguments, each is opened
in a separate tab.  If you don't, CoqIDE opens a single unnamed buffer

> (`*scratch*`).  `coqide` also accepts many of the options of `coqtop`

Author: "**unnamed** buffer **named** (`*scratch*`)" doesn't read very well...

> (`*scratch*`).  `coqide` also accepts many of the options of `coqtop`

Reviewer: How about "titled"?

> (`*scratch*`).  `coqide` also accepts many of the options of `coqtop`

Author: Wait until I push an update and a comment to that effect.  Feedback while I'm in the midst of updating gets really confusing.

> (`*scratch*`).  `coqide` also accepts many of the options of `coqtop`

Author: OK, titled.  I had used "labelled".

(see :ref:`thecoqcommands`), while ignoring the ones that aren't meaningful
for CoqIDE.

.. _coqide_mainscreen:

  .. image:: ../_static/coqide.png
     :alt: CoqIDE main screen

> The screen shot shows CoqIDE while the user is stepping through the file `Fermat.v`.

Author: "Whilst" is an Anglicism.  See https://www.grammarly.com/blog/while-vs-whilst/

A menu bar and tool bar appear at the top of the window. The left-hand panel shows
the current *script buffer*.  Each script buffer corresponds to a separate Coq process.
The upper right panel is the *goal panel*, which shows
the goals to be proven.  The lower right panel has the *message
panel*, which shows messages produced by commands and tactics.

> The contents of the right-hand panels are specific to the currently-displayed script.

Reviewer: There are also tabs for different outputs. Perhaps these should be detailed.

> The status bar is at the bottom.

Reviewer: ```suggestion
The *status bar* is displayed at the bottom.
```

Author: "displayed" isn't really needed, right?

Reviewer: As you wish.

Managing files and buffers, basic editing
-----------------------------------------

The *File* menu lets you open files, create new buffers, save buffers to a file,

> and print or export them into various formats.

Reviewer: ```suggestion
and print or export them in various formats.
```

> Text editing provides the basic operations such as copy/cut/paste and replace.

Reviewer: ```suggestion
Text editing provides the basic operations such as copy, cut, paste, find and replace.
```

Most editing operations are shown in the *Edit* menu, which shows the equivalent

> keystrokes for operations.  if you need more complex editor commands, you can

Reviewer: ```suggestion
keystrokes for operations.  If you need more complex editor commands, you can
```

Perhaps a note about the keystrokes appearing on the right of the operation in a faded font. (Watermark?)

Author: Got it, but above.  I think watermark is not the right word for that.

launch your favorite text editor on the current buffer, using the

> *Edit/External Editor* menu. (When you're done editing you must currently must
> reopen the file to see your changes.)  Also note:

Reviewer: ```suggestion
*Edit/External Editor* menu. When you're done editing you must currently must
reopen the file to see your changes.  Also note:
```

- Undo is also available as Ctrl-Z.  Redo is Ctrl-Shift-Z.
- Select all is Ctrl-A.
- Commenting and uncommenting the current line or selected text is available in

>   the *Tools* menu.

Reviewer: Ctrl (Linux and Windows) and Cmd for Mac I believe. Perhaps we should simply have a note somewhere about Mac users.

Reviewer: Also commenting and commenting is mapped to Ctrl-D and depends on the line of the cursor and the text highlighted.

Author: We should find a Mac user to verify this.  I didn't mention Ctrl-D because the mapping is discoverable from the menu (while Ctrl-Z and Ctrl-A don't appear on any menu).

Author: I think Ctrl-A and Ctrl-Z may be hard coded into the script panel widget.

Per https://support.apple.com/en-us/HT201236, these should be equivalent to Command-A and Command-Z on macos.

FWIW, we need a documented user-friendly way to configure key mappings.  I couldn't get https://github.com/coq/coq/wiki/Configuration-of-CoqIDE to work so far.  You've seen the Edit/Preferences/Shortcuts panel (which works)?  But not in this PR.

Files are automatically saved periodically to a recovery file.  For example,
`foo.v` is saved to `#foo.v#` every 10 seconds by default.  You can change the
interval in the *Edit / Preferences / Files* dialog.

> In the *View* menu, you can set almost a dozen printing options that

Reviewer: 11 to be exact. However I feel as though dozen might be more familiar to native English speakers. Perhaps "several" would work better.

correspond to options that can appear in the script.  For example, "Display
notations" on the menu corresponds to the :flag:`Printing Notations` flag.  You

> should use the menu instead of controlling these settings in your script.

Reviewer: Do we want to link to the Options/Flags section of the refman?

Author: ``` :flag:`Printing Notations` ``` hyperlinks to that specific flag.  I don't think we can link to the flags options index, and the general explanation of flags and options is too basic.  Unfortunately, the menus don't exactly match the flag names, though they are close.

Reviewer: Some of the options set here, set multiple options IIRC. Honestly, I found this entire feature poorly executed, since it also explicitly stops you from setting options within the buffer.

Reviewer: Reported as https://github.com/coq/coq/issues/16080

Running Coq scripts
-------------------

Operations for running the script are available in the *Navigation* menu,
from the toolbar and from the keyboard.  These include:

- Forward (Ctrl-Down) to run one command or tactic
- Backward (Ctrl-Up) undo one command or tactic
- Run to cursor (Ctrl-Right) to run commands up to the cursor
- Run to end (Ctrl-End) to run commands to the end of the buffer
- Reset Coq (Ctrl-Home) to restart the Coq process

> - Interrupt to stop processing commands

Reviewer: End these in periods or no?

Author: No.  They are sentence fragments.  (Yes if they were sentences.)

> Commands may:

Reviewer: Some screenshots may help here?

Author: I added that the toolbar icons have tooltips.  The users can figure it out.  (An exercise for the reader.)  Screenshots take a while to get right, are somewhat system-dependent and quickly get out of date, so I use them sparingly.  For example, compare the first 2 screenshots in the chapter with the third--fairly different.  The first two are old, the third is quite new.

Author: The triage principle applies equally to the documentation--address the most important things first.

- Complete successfully.  In this case, the background of the command is marked
  with the "processed" color (green by default).
- Complete with a warning.  In this case, the warning appears in the messages
  panel in yellow.  The background of the command is marked with the "processed"
  color and the text is shown in blue and underlined.  The message text is
  available as a tooltip on the text of the command.
- Fail with an error.  If you're stepping through the proof line by line, the
  error message appears in the message panel in red and the command is shown
  in red and underlined with a pink background.  If you're in async mode,
  described in more detail later, the message appears in the *errors panel*.
  Double click on an entry to jump to the point of the error.  Execution
  of commands stops unless you're in async mode.

> .. todo: describe the ribbon (a bit broken, though)

Reviewer: How broken? I think the ribbon is a very useful indicator on the progress of a proof and the fact that you can click it to navigate the buffer is quite useful.

Author: I don't know how to explain how to interpret it and its dynamic behavior.  It's animated but doesn't communicate very well.  For example, looking at Pff.v, first I see

![image](https://user-images.githubusercontent.com/1253341/170413436-dcfb689a-b1b3-4b11-b4eb-bcb16f744b67.png)

then a bit later I see

![image](https://user-images.githubusercontent.com/1253341/170413500-02f0fbb5-ac6c-4908-bd54-c60ca21daf36.png)

What do the gray lines mean, eg in Datatypes.v:

![image](https://user-images.githubusercontent.com/1253341/170414203-34cec18b-b664-46c1-b81c-7ffd73d33ebf.png)

It seems to me it's a somewhat unfinished feature.

Reviewer: I would think that `Pff.v` is quite an unusual file due to being over 1000 lines, and I would agree that the ribbon feature in this case is not very helpful. It can be confusing that the ribbon only matches what Coq has actually read, and that the rest of the file is still being read. I don' know if there is an easy way to communicate that.

The grey dots are the unchecked Qeds, and should go away after pressing the check button. I believe this is also not the correct colour. This is probably because the colours in the ribbon are hard coded rather than being derived from the style.

For a 300 line file, for instance. It can be a useful indicator, especially if you have any `Axiom`s or `Admitted` proofs since they show up in yellow.

Reviewer: @jfehrle Would you like to write a little bit about the ribbon however? I think mentioning that it is clickable might be useful.

In the previous figure :ref:`CoqIDE main screen <coqide_mainscreen>`,
the running buffer is `Fermat.v`, all commands until
the ``Theorem`` have been already executed, and the user tried to go
forward executing ``Induction n``. That command failed because no such
tactic exists (names of standard tactics are written in lowercase),
and the failing command is underlined.

> If you're not in async mode and you modify the processed part of the buffer,

Reviewer: Link to async?

Reviewer: Or maybe not it comes quickly after. Perhaps instead we can say, see further down, or something.

Author: I did that on about 10 lines above this.

BTW, do you know how to access the html doc for a specific PR?

Reviewer: Yes, I was reviewing on my phone so I could only read the diff. I will do another pass in HTML.

everything after that point is undone.  Unlike `coqtop`, you should not use

> :cmd:`Undo` to go backward.

Reviewer: Do we ever recommend the use of `Undo` and/or `coqtop` in the refman?

Author: No, IIRC it doesn't work correctly at all.  A couple examples use it, perhaps they should be changed.

Reviewer: I recall vaguely that there was a plan to deprecate and eventually remove it completely. I don't know what happened to that. I would think that the other devs are not particularly fond of keeping it around.

.. todo drop next para, mention left and right with search

There are other buttons on the CoqIDE toolbar: a button to save the running
buffer; a button to close the current buffer (an "X"); buttons to switch among
buffers (left and right arrows); and a "gears" button.

Async mode
----------

Asynchronous mode uses multiple Coq processes to process elements of the script

> in parallel.  For example, multiple theorems may be processed in parallel.  This is

Reviewer: I would say "concurrent" rather than "parallel". Concurrent would mean that the application is processing many separate tasks at the same time, whilst parallel would indicate there is one task broken down into smaller subtasks.

Reviewer: Then again maybe we made a mistake with calling it "parallel proof processing". I suppose the task depends on your point of view.

Author: It is parallel (when things go well :-) ).  Coq creates multiple proof worker processes that can run on different cores.  Not merely concurrent.

Reviewer: Right, but my point was that they are not processing the same proof, but separate proofs in the same buffer. Minor point anyway.

Author: The "multiple theorems" can be processed in parallel, but individual theorems can't be split and parallelized as things are now. (Maybe different in the future).  I think that would be the most obvious understanding from the current wording.

described in detail in :ref:`asynchronousandparallelproofprocessing`.

While synchronous mode stops processing at the first error it encounters, in async

> mode, errors only stop processing the proof the error appears in.

Reviewer: Do we want to mention that a "proof" is delaminated by `Proof.` and `Qed.`? I am not even sure `Proof.` is even needed.

Author: Got it. I think you mean "delimited", eh?  I was told `Proof.` was required.  If I say it's required and it's not, I doubt anyone will be confused or will complain.

Reviewer: `Proof` is recommended but not required. This will likely change in the future, but there is still some pushback from users that do not want to write 6 characters.

Therefore it can report errors in multiple proofs without manual intervention.
In addition, async mode lets the user edit failed proofs without invalidating
successful proofs that appear after it in the script.

.. image:: ../_static/async-mode.png
   :alt: Async mode

> In the screenshot, the proof of the failed theorem can be edited (between `Proof.`
> and `Qed.`) without invalidating the theorems that follow it.  Then the modified
> proof can be reprocessed using the usual navigation operations.  The light blue
> highlight in the script indicates commands that haven't been processed.

Reviewer: ```suggestion
In the screenshot, the proof of the failed theorem can be edited (between `Proof.`
and `Qed.`) without invalidating the theorems that follow it.  The modified
proof can then be reprocessed using the usual navigation operations.  The light blue
highlight in the script indicates commands that haven't been processed.
```

Async mode defers the final type checking step of proofs, leaving the `Qed.` marked in
a slightly different shade of light blue to indicate this.  To complete the final

> checking, click on the "gears" button on the toolbar ("Fully check the document").

Reviewer: Wow, I only just now found out about this.

Commands and templates
----------------------

The Templates menu allows using shortcuts to insert
commands. This is a nice way to proceed if you're not sure of the
syntax of the command you want.

Moreover, from this menu you can automatically insert templates of complex
commands like ``Fixpoint`` that you can conveniently fill afterwards.

Queries
-------

.. image:: ../_static/coqide-queries.png
   :alt: CoqIDE queries

A *query* is any command that does not change the current state, such as
:cmd:`Check`, :cmd:`Search`, etc.  The *query pane* lets you run such commands
interactively without modifying your script. The query pane is accesible from
the *View* menu, or using the shortcut ``F1``.

> You can also do queries by selecting a particular phrase, then choosing an
> item from the *Queries* menu. The response will appear in the message panel.

Reviewer: ```suggestion
You can also do queries by selecting a particular phrase, then choosing an
item from the *Queries* menu. The response will appear in the message panel.
```

Selecting a phrase or highlighting text with the cursor? Phrase doesn't have any technical meaning IIRC.

Coq Integrated Development Environment
========================================

.. todo PR: how to say that a number of things are broken?  Maybe list them
   somewhere--doesn't have to be super detailed

> The Coq Integrated Development Environment (CoqIDE) is a user-friendly GUI
> for Coq. Its main purpose is to
> allow users to edit scripts and step forward and backward through Coq scripts,
> respectively executing and undoing commands and tactics.

Reviewer: This paragraph is hard to read IMO.  
```suggestion
The Coq Integrated Development Environment (CoqIDE) is a user-friendly GUI
for Coq. Its main purpose is to
allow users to edit scripts and step through Coq scripts, executing and undoing commands and tactics.
```

The respectively part encourages me to reread, but who knows, maybe I can't read.

Reviewer: Also please rehard wrap.

Author: Rewrapping is a manual process.  My GUI doesn't have a command to do that.  Also it doesn't matter, does it?  We expect users to read the HTML.  But I did rewrap this couple of lines.

Reviewer: @jfehrle My GUI allows me to do some rewrapping, it can help when reading the rst files directly, which is sometimes done. It also helps people when they edit it later. Not a big deal if you cannot do it right now however.

Author: It's OK to rewrap the changed parts, but en masse rewrapping would obscure the git history.  I expect this is not news to you.

To run CoqIDE, enter `coqide` on the command line.

> If you include script file names as arguments, each is opened

Reviewer: Do we mention that "script file" = ".v file"? We should perhaps include (.v) or something.

Author: I'm sure it's in the document somewhere but some repetition is fine.

in a separate tab.  If you don't, CoqIDE opens a single unnamed buffer
(titled `*scratch*`).  `coqide` also accepts many of the options of `coqtop`
(see :ref:`thecoqcommands`), while ignoring the ones that aren't meaningful
for CoqIDE.  Use `coqide --help` to see the list of command line options.

.. _coqide_mainscreen:

  .. image:: ../_static/coqide.png
     :alt: CoqIDE main screen

> The screen shot shows CoqIDE while the user is stepping through the file `Fermat.v`.

Reviewer: ```suggestion
The screen shot shows the interface of CoqIDE as the user is stepping through the file `Fermat.v`.
```

Author: Does "the interface of" add really anything?  I think it just makes the text wordier.  I'm a big fan of being succinct (subject to also being clear).

Reviewer: OK then don't worry about the first part.

> A menu bar and tool bar appear at the top of the window. The left-hand panel shows

Reviewer: ```suggestion
A menu bar and a tool bar appear at the top of the window. The left-hand panel shows
```
I'm not sure about this one however. But I'll state it in case you agree.

the current *script buffer*.  Each script buffer corresponds to a separate Coq process.
The upper right panel is the *goal panel*, which shows
the goals to be proven.  The lower right panel has three tabs: the *Messages
panel*, which shows messages produced by commands and tactics; the *Errors panel*,
which shows errors detected when running in :ref:`async mode <asyncmode>` and the *Jobs panel,* which shows
information on the worker processes used by async mode.

> The contents of the right-hand panels are specific to the currently-displayed script.

Reviewer: We should also mention the "pop-out" feature, which lets you pop the tabs out in their seperate window, to be restored when the window is closed.

Author: We call it "detached" (as in `Windows/Detach proof` on the menu and the `Detach` buttons on the Query panel and debugger panels.

Reviewer: Yep, detached is the correct word. Will you mention it here then?

Author: Yup.

The *status bar* is a line of text that appears at the bottom of the window.

Managing files and buffers, basic editing
-----------------------------------------

The *File* menu lets you open files, create new buffers, save buffers to a file,
and print or export them in various formats.

Text editing provides the basic operations such as copy, cut, paste, find and replace.
Most editing operations are shown in the *Edit* menu.  Most menu entries show the equivalent
keystrokes to perform the action.  If you need more complex editor commands, you can

> launch your favorite text editor on the current buffer, using the

Reviewer: Your "favorite text editor" is configurable in the preferences menu.

Author: While `Edit/External editor` saves the buffer if it's modified, IMHO when that editor exits, CoqIDE should reload the file if has been changed.  We also should have a way for CoqIDE to detect and act on external changes to .v files, .vo files, etc.

The external editor fails if you're running in from the source tree.  Also there should be a message if the editor doesn't come up.

*Edit/External Editor* menu. When you're done editing, you must currently must
reopen the file to see your changes.  Also note:

- Undo is also available as Ctrl-Z (Command-Z on macOS).  Redo is Ctrl-Shift-Z

>   (Shift-Command-Z on macOS).

Reviewer: MacOS shortcuts are typically abbreviated to Cmd rather than Command. The same way Control is abbreviated to Ctrl on Windows and Linux. Mac keyboards (that I have seen) have Cmd written on them.

Author: The Apple web page I consulted had it spelled out.  But "Cmd" is an abbreviation similar to "Ctrl" and "Alt", so more appropriate.

Reviewer: Here is how Apple shortens it:

![image](https://user-images.githubusercontent.com/8614547/170846231-9faf5881-666a-4add-b947-3b978dece1e5.png)

- Select all is Ctrl-A (Command-A on macOS).
- Ctrl-Delete deletes a word of text after the cursor

> - Ctrl-Backspace deletes a word of text before the cursor

Reviewer: Perhaps for these keyboard shortcuts, we should use code blocks to better distinguish from the text.

Author: Done.  I think we need a font that's more distinctive (later on):

![image](https://user-images.githubusercontent.com/1253341/170846029-9534ba85-74a0-49f3-acd0-d3657bff66ef.png)

  the *Tools* menu.  If some text is selected, exactly that text is commented out;
  otherwise the line containing the cursor is commented out.  To uncomment, position
  the cursor between `(*` and `*)` or select any text between them.

Files are automatically saved periodically to a recovery file.  For example,
`foo.v` is saved to `#foo.v#` every 10 seconds by default.  You can change the
interval in the *Edit / Preferences / Files* dialog.

In the *View* menu, you can set several printing options that
correspond to options that can appear in the script.  For example, "Display
notations" on the menu corresponds to the :flag:`Printing Notations` flag.  You
should use the menu instead of controlling these settings in your script.

Running Coq scripts
-------------------

Operations for running the script are available in the *Navigation* menu,
from the toolbar and from the keyboard.  These include:

- Forward (Ctrl-Down) to run one command or tactic
- Backward (Ctrl-Up) undo one command or tactic
- Run to cursor (Ctrl-Right) to run commands up to the cursor
- Run to end (Ctrl-End) to run commands to the end of the buffer
- Reset Coq (Ctrl-Home) to restart the Coq process

> - Interrupt to stop processing commands

Reviewer: In my experience, if CoqIDE is very busy, for example printing a huge term, interrupt can be somewhat unresponsive. The goal is to not undo the processed part of the script and just interrupt the current command. If CoqIDE is really not responsive however, resetting Coq with Ctrl-Home typically does the job.

I don't know if this needs mentioning.

Author: Interrupt is implemented with a signal sent directly to coqidetop that sets a flag.  The interpreter checks the flag before starting the next command.  I tweaked the wording.

Tooltips identify the action assocaited with each toolbar icon.

Commands may:

- Complete successfully.  In this case, the background of the command is marked
  with the "processed" color (green by default).
- Complete with a warning.  In this case, the warning appears in the messages
  panel in yellow.  The background of the command is marked with the "processed"
  color and the text is shown in blue and underlined.  The message text is
  available as a tooltip on the text of the command.
- Fail with an error.  If you're stepping through the proof line by line, the
  error message appears in the message panel in red and the command is shown
  in red and underlined with a pink background.  If you're in async mode,
  described in more detail below, the message appears in the *errors panel*.
  Double click on an entry to jump to the point of the error.  Execution

>   of commands stops unless you're in async mode.

Reviewer: Also here, we have Axiom's and Admitted which are highlighted in yellow in the buffer.

Author: Think I'd call it light orange.  Different shades in the script and in the ribbon BTW.  Done. but OTOH this is something that users will effortlessly observe (I don't recall ever being told about this behavior).

.. todo: describe the ribbon (a bit broken, though)

In the previous figure :ref:`CoqIDE main screen <coqide_mainscreen>`,
the running buffer is `Fermat.v`, all commands until

> the ``Theorem`` have been already executed, and the user tried to go

Reviewer: ```suggestion
the ``Theorem`` have already been executed, and the user has tried to go
```

Author: How about "all commands until the ``Theorem`` have already been executed, then the user tried to go"?

> forward executing ``Induction n``. That command failed because no such

Reviewer: ```suggestion
forward executing ``Induction n``. That command has failed, since no such
```

Author: I think the simple past is fine.  You don't like "because" here?

Reviewer: OK fair enough, I think yours is simpler.

tactic exists (names of standard tactics are written in lowercase),

> and the failing command is underlined.

Reviewer: ```suggestion
and the failing command has been underlined.
```

Therefore it can report errors in multiple proofs without manual intervention.
In addition, async mode lets the user edit failed proofs without invalidating
successful proofs that appear after it in the script.  The editable part of the proof
is the part that appears between `Proof.` and `Qed.`.

.. image:: ../_static/async-mode.png
   :alt: Async mode

In the screenshot, the proof of the failed theorem can be edited (between `Proof.`
and `Qed.`) without invalidating the theorems that follow it.  The modified
proof can then be reprocessed using the usual navigation operations.  The light blue
highlight in the script indicates commands that haven't been processed.

Async mode defers the final type checking step of proofs, leaving the `Qed.` marked in
a slightly different shade of light blue to indicate this.  To complete the final
checking, click on the "gears" button on the toolbar ("Fully check the document").

Commands and templates
----------------------

The Templates menu allows using shortcuts to insert
commands. This is a nice way to proceed if you're not sure of the
syntax of the command you want.

Moreover, from this menu you can automatically insert templates of complex

> commands like ``Fixpoint`` that you can conveniently fill afterwards.

Reviewer: ```suggestion
commands like ``Fixpoint`` that you can conveniently fill in afterwards.
```

Queries
-------

.. image:: ../_static/coqide-queries.png
   :alt: CoqIDE queries

A *query* is any command that does not change the current state, such as

> :cmd:`Check`, :cmd:`Search`, etc.  The *query pane* lets you run such commands

Reviewer: ```suggestion
:cmd:`About`, :cmd:`Check`, `cmd:`Print`, :cmd:`Search`, etc.  The *query pane* lets you run such commands
```

Author: Got it.  Do we still need the "etc."?

Reviewer: Technically there are a few more so I would keep it.

interactively without modifying your script. The query pane is accessible from
the *View* menu, or using the shortcut ``F1``.
You can also do queries by selecting some text, then choosing an
item from the *Queries* menu. The response will appear in the message panel.
The image above shows the result after selecting
``Nat.mul`` in the bottom line of the script panel, then choosing ``Print``
from the ``Queries`` menu.

.. todo: should names of menus be *Menu* or `Menu` or ??  not consistent

Compilation
-----------

> The `Compile` menu offers direct commands to:

Reviewer: Shall we mention that this doesn't work at the moment, might get fixed in the future and that compiling from the terminal is a more sound alternative?

Author: In this case, yes.  It's on my list to fix. I expect it will be straightforward and that I'll get to it fairly soon.

+ create a `Makefile` using `coq_makefile`.

Customizations
--------------

Preferences
~~~~~~~~~~~

You may customize your environment with the *Preferences* dialog, which

> *Edit/Preferences* on the menu. There are several sections:

Reviewer: Which Edit/Preferences? Do you mean "which is found under Edit/Preferences"?

     ; (gtk_accel_path "<Actions>/Export/Export to" "")
     (gtk_accel_path "<Actions>/Edit/Find Next" "F4")

The first line corresponds to the menu item for the Queries/About menu item,
which was bound by default to Shift-Ctrl-A.  "<Primary>" indicates Ctrl.
The second line is for a menu item that has no key binding.

Lines that begin with semicolons are comments create by CoqIDE.  CoqIDE uses
the default binding
for these items.  To change a key binding, remove the semicolon and set the third item
in the list as desired, such as in the third line.  Avoid assigning the same binding
to multiple items.

If the same menu item name appears on multiple lines in the file, the value from the
last line is used.  This is convenient for copying a group of changes from elsewhere--just
insert the changes at the end of the file.  The next time CoqIDE terminates, it will
usually reorder the items.

Some menu entries can be changed as a group from the Edit/Preferences/Shortcuts panel.
Key bindings that don't appear in the file such as Ctrl-A (Select All) can't be
changed with this mechanism.

.. todo: list common rebindings?

> .. todo: microPG mode?

Reviewer: I don't particularly remember, but I seem to be under the impression this enables the PG keybindings rather than the default CoqIDE ones.

Author: Yeah, I've not found time to experiment with it.  See `Help/Help for uPG mode` and `Preferences/Editor`.  On the former, I wonder what key `Meta` corresponds to.  Also what happens to the accelerator settings it clashes with.
